### PR TITLE
feat: support maxLength prop in web

### PR DESCRIPTION
**Pull request by dominictb**
### Details

This PR includes the implementation of `maxLength` prop in web

### Related Issues

https://github.com/Expensify/App/issues/46766

### Manual Tests

1. Paste this text file (over 400k characters) 
[default.txt](https://github.com/user-attachments/files/17113641/default.txt)

2. Check that the text input in both web and native is cut off to 30k length

### Linked PRs

Pending App PR

### CODE REVIEW

**Comment by Skalakid** [comment]
Hello, I'm testing your changes and I found couple of bugs:

1. The input `value` isn't matching the input content. The `value` (in the picture the text under the input) should be the same as the text inside the component

Screenshot 2024-09-26 at 12 14 49

**Comment by Skalakid** [comment]
2. There are some problems with entering text when we reach the `maxLength` limit. Steps:
 - move the cursor to the end of the text (on the `maxLength` index)
 - start typing
 - notice that the old text is being replaced

https://github.com/user-attachments/assets/3dca11e3-60f3-483e-a366-0cc2fc62fe5f

**Comment by Skalakid** [comment]
Okay, thank you. Now everything works fine :D 

There is one more thing that I wanted to fix there. How about adding value cutting on the first rerender? Currently, when you pass the long text in the `value` prop and the app loads, the `maxLength` prop isn't cutting the input value:

Screenshot 2024-09-26 at 13 22 47

(On the image, I have `maxLength={5}` and default text passed in the `value` prop)

I know the same thing happens on the default RN TextInput, but I think it's worth fixing in our case. What do you think about it?

**Comment by Skalakid** [comment]
Bug: Text isn't being replaced when you reach the limit of `maxLength` prop

https://github.com/user-attachments/assets/a1a7b3fe-05bd-4955-a6ef-b65fe769d520

Expected behavior:


https://github.com/user-attachments/assets/5321189d-80c6-422a-a0a9-5f3330995c94

File: src/MarkdownTextInput.web.tsx [hunk 1]
```diff
@@ -305,14 +316,18 @@ const MarkdownTextInput = React.forwardRef<TextInput, MarkdownTextInputProps>(
...
-        const parsedText = normalizeValue(
+        let parsedText = normalizeValue(
           inputType === 'pasteText' ? pasteContent.current || '' : parseInnerHTMLToText(e.target as MarkdownTextInputElement, inputType, contentSelection.current.start),
         );
 
         if (pasteContent.current) {
           pasteContent.current = null;
         }
 
+        if (typeof maxLength === 'number' && parsedText.length > maxLength) {
```
Comment (Skalakid) [inline src/MarkdownTextInput.web.tsx:327]:
```suggestion
        if (maxLength !== undefined && parsedText.length > maxLength) {
```

File: src/MarkdownTextInput.web.tsx [hunk 2]
```diff
@@ -400,17 +414,25 @@ const MarkdownTextInput = React.forwardRef<TextInput, MarkdownTextInputProps>(
         }
 
         const previousText = divRef.current.value;
-        const newText = `${divRef.current.value.substring(0, contentSelection.current.start)}${text}${divRef.current.value.substring(contentSelection.current.end)}`;
+        let insertedText = text;
+        let availableLength = text.length;
+        const prefix = divRef.current.value.substring(0, contentSelection.current.start);
+        const suffix = divRef.current.value.substring(contentSelection.current.end);
+        if (typeof maxLength === 'number') {
```
Comment (Skalakid) [inline src/MarkdownTextInput.web.tsx:421]: Same here

File: src/MarkdownTextInput.tsx [hunk 3]
```diff
@@ -36,15 +36,32 @@ function processMarkdownStyle(input: PartialMarkdownStyle | undefined): Markdown
...
 }
 
-const MarkdownTextInput = React.forwardRef<TextInput, MarkdownTextInputProps>((props, ref) => {
+const MarkdownTextInput = React.forwardRef<TextInput, MarkdownTextInputProps>(({value, maxLength, onChangeText, ...props}, ref) => {
   const IS_FABRIC = 'nativeFabricUIManager' in global;
 
   const markdownStyle = React.useMemo(() => processMarkdownStyle(props.markdownStyle), [props.markdownStyle]);
 
+  const processedText = React.useMemo(() => {
+    if (typeof maxLength === 'number') {
```
Comment (Skalakid) [inline src/MarkdownTextInput.tsx:45]:
```suggestion
    if (maxLength === undefined) {
```

File: src/MarkdownTextInput.tsx [hunk 4]
```diff
@@ -36,15 +36,32 @@ function processMarkdownStyle(input: PartialMarkdownStyle | undefined): Markdown
...
+      return value?.slice(0, maxLength);
+    }
+    return value;
+  }, [value, maxLength]);
+
+  React.useLayoutEffect(() => {
+    if (typeof maxLength !== 'number' || !value || value.length <= maxLength || !onChangeText) {
```
Comment (Skalakid) [inline src/MarkdownTextInput.tsx:52]:
```suggestion
    if (maxLength === undefined || !value || value.length <= maxLength || !onChangeText) {
```

**Review by Skalakid — COMMENTED** [review]
Overall, it looks good :D Left some small comments

**Review by Skalakid — APPROVED** [review]
LGTM

File: src/MarkdownTextInput.tsx [hunk 5]
```diff
@@ -36,15 +36,35 @@ function processMarkdownStyle(input: PartialMarkdownStyle | undefined): Markdown
...
+    }
+    onChangeTextRef.current(value.slice(0, maxLength));
+  }, [value, maxLength]);
+
   return (
     <>
       <TextInput
         {...props}
+        value={processedText}
+        onChangeText={onChangeText}
```
Comment (tomekzaw) [inline src/MarkdownTextInput.tsx:66]: What if the component was supposed to be uncontrolled? I don't really like this logic here

**Review by tomekzaw — CHANGES_REQUESTED** [review]

**Comment by tomekzaw** [comment]
@dominictb Why can't we just pass `maxLength` prop to `TextInput` in the native implementation?

**Comment by tomekzaw** [comment]
@dominictb Thanks for the response

> Basically for value > maxLength, native component will just render the whole value without cutting off anything

Isn't it something we should fix directly in react-native?

**Comment by tomekzaw** [comment]
Actually, a regular `<input type="text">` on web also behaves the same way as currently – if you set `maxlength="10" value="1234567890abcde"`, all 15 characters are visible but you can't edit the text field. I think this is the expected behavior.

**Comment by Skalakid** [comment]
Okay, @tomekzaw convinced me, let's revert changes on the native side and focus on implementing the `maxLength` prop on the web only. Let's make sure it will work the same as the native one and isn't introducing any new bug

**Comment by tomekzaw** [comment]
Basically, `maxLength` prop for `MarkdownTextInput` should work the same way as for `TextInput`.

For Android and iOS, let's just pass this prop to `TextInput` just like we already do (via `{...props}`).

For web, we'll need to implement `maxLength` logic since we're using `contenteditable` which doesn't support `maxlength` out-of-the-box.

All other logic should be moved to E/App.

**Comment by tomekzaw** [comment]
@dominictb Thanks!

**Review by Skalakid — DISMISSED** [review]
LGTM

**Review by tomekzaw — APPROVED** [review]

**Pull request merged**